subsys/bluetooth/nrf_bt_adv/core.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bluetooth/nrf_bt_adv.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nrf_bt_adv, CONFIG_NRF_BT_ADV_LOG_LEVEL);
+
+enum provider_set {
+	PROVIDER_SET_AD,
+	PROVIDER_SET_SD
+};
+
+static void get_section_ptrs(enum provider_set set,
+			     const struct nrf_bt_adv_data_provider **start,
+			     const struct nrf_bt_adv_data_provider **end)
+{
+	switch (set) {
+	case PROVIDER_SET_AD:
+		{
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_ad_list_start[];
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_ad_list_end[];
+
+			*start = _nrf_bt_adv_ad_list_start;
+			*end = _nrf_bt_adv_ad_list_end;
+		}
+		break;
+
+	case PROVIDER_SET_SD:
+		{
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_sd_list_start[];
+			extern const struct nrf_bt_adv_data_provider _nrf_bt_adv_sd_list_end[];
+
+			*start = _nrf_bt_adv_sd_list_start;
+			*end = _nrf_bt_adv_sd_list_end;
+		}
+		break;
+
+	default:
+		/* Should not happen. */
+		__ASSERT_NO_MSG(false);
+		*start = NULL;
+		*end = NULL;
+		break;
+	}
+}
+
+static size_t get_section_size(enum provider_set set)
+{
+	const struct nrf_bt_adv_data_provider *start;
+	const struct nrf_bt_adv_data_provider *end;
+
+	get_section_ptrs(set, &start, &end);
+
+	return end - start;
```

Comment:
Shouldn't it be divided by 4? (each provider section takes 4 bytes)

Reply:
I think not. We subtract pointers of type: `const struct nrf_bt_adv_data_provider *` (not uint8_t pointers)

Reply:
Right